sirepo/pkcli/runner.py

```diff
@@ -0,0 +1,366 @@
+                        _write_status(runner_client.JobStatus.ERROR, run_dir)
+                    else:
+                        _write_status(runner_client.JobStatus.COMPLETED, run_dir)
+            finally:
+                # _write_status is a no-op if there's already a status, so
+                # this really means "if we get here without having written a
+                # status, assume there was some error"
+                _write_status(runner_client.JobStatus.ERROR, run_dir)
+                # Make sure that we clear out the running job info and tell
+                # everyone the job is done, no matter what happened
+                job_info = self.jobs.pop(run_dir, None)
+                if job_info is not None:
+                    job_info.finished.set()
+
+
+_RPC_HANDLERS = {}
+
+
+def _rpc_handler(fn):
+    _RPC_HANDLERS[fn.__name__.lstrip('_')] = fn
+    return fn
+
+
+@_rpc_handler
+async def _start_job(job_tracker, request):
+    pkdc('start_job: {}', request)
+    await job_tracker.start_job(
+        request.run_dir, request.jhash, request.cmd
+    )
+    return {}
+
+
+@_rpc_handler
+async def _job_status(job_tracker, request):
+    pkdc('job_status: {}', request)
+    return {
+        'status': job_tracker.status(request.run_dir, request.jhash).value
+    }
+
+
+@_rpc_handler
+async def _cancel_job(job_tracker, request):
+    if request.run_dir not in job_tracker.jobs:
+        return {}
+    job_info = job_tracker.jobs[request.run_dir]
+    if job_info.status is not runner_client.JobStatus.RUNNING:
+        return {}
+    job_info.status = runner_client.JobStatus.CANCELED
+    _write_status(runner_client.JobStatus.CANCELED, request.run_dir)
+    job_info.process.terminate()
+    with trio.move_on_after(_KILL_TIMEOUT_SECS):
+        await job_info.finished.wait()
+    job_info.process.kill()
+    await job_info.finished.wait()
+    return {}
+
+
+# XX should we just always acquire a per-job lock here, to make sure we never
+# have to worry about different requests for the same job racing?
+async def _handle_conn(job_tracker, lock_dict, stream):
+    with _catch_and_log_errors(Exception, 'error handling request'):
+        try:
+            request_bytes = bytearray()
+            while True:
+                chunk = await stream.receive_some(_CHUNK_SIZE)
+                if not chunk:
+                    break
+                request_bytes += chunk
+            request = pkjson.load_any(request_bytes)
+            if 'run_dir' in request:
+                request.run_dir = pkio.py_path(request.run_dir)
+            pkdp('runner request: {!r}', request)
+            handler = _RPC_HANDLERS[request.action]
+            async with lock_dict[request.run_dir]:
+                response = await handler(job_tracker, request)
+            pkdp('runner response: {!r}', response)
+            response_bytes = pkjson.dump_bytes(response)
+        except Exception as exc:
+            await stream.send_all(
```

Comment:
If there is an error on the stream itself, you can't send to it. It's safer to protect against that rather than raise another exception. 

I think it shouldn't send back the error string, and instead, just send back "error". It breaks the abstraction to respond with raw errors. Rather, just log the error here. If there's a known error, translate it to a known error code for the sender.

Logging might look like: `pkdlog('Request={} error={} stack=', request, exc, pkdexc())`

Reply:
> If there is an error on the stream itself, you can't send to it. It's safer to protect against that rather than raise another exception.

In Python 3 we have implicit exception chaining, so if `stream.send_all` raises an exception, then both the original exception+traceback and the new exception+traceback will both be logged.

Also, we're already logging any leaking exceptions with `_catch_and_log_errors`, so that's not an issue.

Easy enough to just send back "error" though, I'll do that.

Reply:
Actually, we might as well delete this bit and make the error indicator be "empty response" – adding fancier error handling code just produces more chances for things to go wrong :-)

Reply:
> In Python 3 we have implicit exception chaining, so if `stream.send_all` raises an exception, then both the original exception+traceback and the new exception+traceback will both be logged.

...actually, I'm wrong, because `pkdexc` is buggy on py3 :-( https://github.com/radiasoft/pykern/issues/29

Reply:
> > If there is an error on the stream itself, you can't send to it. It's safer to protect against that rather than raise another exception.
> 
> Easy enough to just send back "error" though, I'll do that.

To be clearer: `stream` is in error, you shouldn't send on it, because you'll get a cascade that could be avoided. I feel like in error code, we have to take pains to avoid that type of cascade, because it makes debugging harder, even if py3 gives you all the tracebacks.